File: Samples/JSSample/src/Protocol/ShipMsgCaller.ts
import { CSmartValue, CSmartArray, CSmartStruct } from "../DOSSystem/SmartStruct";
import { IDistributedObjectOperator, COMMON_RESULT_CODE, DOSMakeMsgID, OBJECT_ID } from "../DOSSystem/DOSSystem";
import { MODULE_IDS, DOSSAMPLE_INTERFACES } from "./DOSSampleSystemEnums";
import { IShip } from "./ShipInterface";
import { SHIP_METHODS } from "./ShipInterface";
import { SHIP_ATTACK_MEMBER_IDS } from "./ShipInterface"


export class CShipMsgCaller implements IShip	
{
	m_Operator: IDistributedObjectOperator;
	m_SenderID: OBJECT_ID;
	m_TargetID: Array<OBJECT_ID>;
		
	constructor(Operator: IDistributedObjectOperator, SenderID: OBJECT_ID, TargetID: OBJECT_ID | Array<OBJECT_ID>)
	{
		this.m_Operator = Operator;
		this.m_SenderID = SenderID;
		if (TargetID instanceof Array)
			this.m_TargetID = TargetID;
		else
			this.m_TargetID = [TargetID];
	}
		
		
		
	public Attack(Damage: number ): number
	{
		if((this.m_Operator == null) || (this.m_Operator == undefined))
		{
			return COMMON_RESULT_CODE.COMMON_RESULT_FAILED;
		}
	
		let _PackSize = CSmartStruct.GetEmptyStructSize();
		_PackSize += CSmartStruct.GetFixMemberSize(4);
		
		
	
		let MsgPacket = this.m_Operator.NewMessagePacket(_PackSize, this.m_TargetID.length);	
		MsgPacket.SenderID = this.m_SenderID;
		let MsgFlag=0;
		MsgPacket.MsgFlag = MsgFlag;
		
	    MsgPacket.MsgID = DOSMakeMsgID(MODULE_IDS.MODULE_ID_DOSSAMPLE,DOSSAMPLE_INTERFACES.DOSSAMPLE_INTERFACE_SHIP,SHIP_METHODS.METHOD_ATTACK,false);		
	
		
		let Packet = MsgPacket.EmptyDataPacket;
		
		if(this.PackMsgAttack(Packet,Damage))
		{
			MsgPacket.DataLength = Packet.GetDataLen();
			MsgPacket.TargetIDs = this.m_TargetID;
			if(this.m_Operator.SendMessagePacket(MsgPacket))
			{
				return COMMON_RESULT_CODE.COMMON_RESULT_SUCCEED;
			}
			else
			{
				return COMMON_RESULT_CODE.COMMON_RESULT_MSG_SEND_ERROR;
			}
		}
		else
		{
			return COMMON_RESULT_CODE.COMMON_RESULT_MSG_PACK_ERROR;
		}
		
	}
	
	
	public PackMsgAttack(Packet: CSmartStruct,Damage: number ): boolean
	{
		let FailCount=0;
	
		{
			if(!Packet.AddMemberInt(SHIP_ATTACK_MEMBER_IDS.SST_ATTACK_DAMAGE,Damage)) FailCount++;
		}
		
	
		return (FailCount==0);
	}
	
	
};

File: Samples/CSSample/Procotol/ShipMsgCaller.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using DOSSystem;
using DOSSample;


namespace DOSSample
{
	public class CShipMsgCaller :
		IShip
	{
		DistributedObjectOperator						m_Operator;		
		OBJECT_ID										m_SenderID;
		OBJECT_ID[]										m_TargetIDList;
		
		public CShipMsgCaller(DistributedObjectOperator Operator, OBJECT_ID SenderID, OBJECT_ID TargetID)
		{
			m_Operator = Operator;
			m_SenderID = SenderID;
			m_TargetIDList = new OBJECT_ID[1]{TargetID};
		}
		public CShipMsgCaller(DistributedObjectOperator Operator, OBJECT_ID SenderID, OBJECT_ID[] TargetIDList)
		{
			m_Operator = Operator;
			m_SenderID = SenderID;
			m_TargetIDList = TargetIDList;
		}
		
		
		public int Attack(int Damage )
		{
			if(m_Operator == null)
			{
				return (int)COMMON_RESULT_CODE.COMMON_RESULT_FAILED;
			}
		
			uint _PackSize = CSmartStruct.GetEmptyStructSize();
			_PackSize += CSmartStruct.GetFixMemberSize(sizeof(int));
			
			;
		
			ushort MsgFlag = 0;
				
			byte[] MsgBuffer=new byte[_PackSize];
			
			
		    uint MsgID = DistributedObject.MakeMsgID((uint)MODULE_IDS.MODULE_ID_DOSSAMPLE,(uint)DOSSAMPLE_INTERFACES.DOSSAMPLE_INTERFACE_SHIP,(uint)SHIP_METHODS.METHOD_ATTACK,false);		
		
			
			CSmartStruct Packet = new CSmartStruct(MsgBuffer,0,_PackSize,true);
			
			if(PackMsgAttack(Packet,Damage))
			{	
				if(m_Operator.SendMessageMulti(m_TargetIDList, true, m_SenderID, MsgID, MsgFlag, MsgBuffer, 0, MsgBuffer.Length))
				{
					return (int)COMMON_RESULT_CODE.COMMON_RESULT_SUCCEED;
				}
				else
				{
					return (int)COMMON_RESULT_CODE.COMMON_RESULT_MSG_SEND_ERROR;
				}
			}
			else
			{
				return (int)COMMON_RESULT_CODE.COMMON_RESULT_MSG_PACK_ERROR;
			}
			
		}
		
		
		public bool PackMsgAttack(CSmartStruct Packet,int Damage )
		{
			uint FailCount=0;
		
			{
				if(!Packet.AddMember((ushort)SHIP_ATTACK_MEMBER_IDS.SST_ATTACK_DAMAGE,Damage)) FailCount++;
			}
			
			
		
			return (FailCount==0);
		}
		
		
	};
};
